apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common.Tests/Trace/TraceHeaderPropagatingHandlerTest.cs

```diff
@@ -0,0 +1,129 @@
+            mockTracer.Verify(t => t.EndSpan(), Times.Never());
+        }
+
+        [Fact]
+        public async Task SendAsync_Trace()
+        {
+
+            var mockTracer = GetSetUpTracer();
+            var traceHandler = TraceHeaderPropagatingHandler.Create(mockTracer.Object);
+            traceHandler.InnerHandler = new FakeDelegatingHandler(headerContext);
+
+            var requestUri = new Uri("https://www.google.com");
+
+            using (var httpClient = new HttpClient(traceHandler))
+            {
+                await httpClient.GetAsync(requestUri);
+            }
+
+            mockTracer.Verify(t => t.StartSpan(requestUri.ToString(), null), Times.Once());
+            mockTracer.Verify(t => t.EndSpan(), Times.Once());
+        }
+
+        [Fact]
+        public async Task SendAsync_TraceException()
+        {
+            var mockTracer = GetSetUpTracer();
+            var traceHandler = TraceHeaderPropagatingHandler.Create(mockTracer.Object);
+            traceHandler.InnerHandler = new FakeDelegatingHandler(headerContext, true);
+
+            var requestUri = new Uri("https://www.google.com");
+
+            using (var httpClient = new HttpClient(traceHandler))
+            {
+                await Assert.ThrowsAsync<Exception>(() => httpClient.GetAsync(requestUri));
+            }
+
+            mockTracer.Verify(t => t.StartSpan(requestUri.ToString(), null), Times.Once());
+            mockTracer.Verify(t => t.SetStackTrace(It.IsAny<StackTrace>()), Times.Once());
+            mockTracer.Verify(t => t.EndSpan(), Times.Once());
+        }
+
+        /// <summary>
+        /// Fake <see cref="DelegatingHandler"/> to verify added trace headers.
+        /// </summary>
+        private class FakeDelegatingHandler : DelegatingHandler
+        {
+            private readonly TraceHeaderContext _context;
+            private readonly bool _throwException;
+
+            public FakeDelegatingHandler(TraceHeaderContext context = null, bool throwException = false)
+            {
+                _context = context;
+                _throwException = throwException;
+            }
+
+            protected override Task<HttpResponseMessage> SendAsync(
+                HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                if (_context != null)
+                {
+                    var traceHeader = request.Headers.GetValues(TraceHeaderContext.TraceHeader).First();
+                    var currentContext = TraceHeaderContext.FromHeader(traceHeader);
+                    Assert.Equal(_context.TraceId, currentContext.TraceId);
+                    Assert.Equal(_context.SpanId, currentContext.SpanId);
+                    Assert.Equal(_context.ShouldTrace, currentContext.ShouldTrace);
+                }
+                if (_throwException)
+                {
+                    throw new Exception();
```

Comment:
I think this might cause a problem for any test which is expecting an exception. It will `Assert.Throws` on type `Exception`, but if any of the 3 asserts above fail, that expectation will actually be satisfied and the test will pass (I could be wrong though - I'm not sure if `Assert.Throws` checks for type equality or a type being an instance of the expected one, but I assume the latter). You should throw (and expect) some other exception type that is not in the type hierarchy of whatever `Assert.Equal` throws.